logic/Rule.java

```diff
@@ -472,42 +472,118 @@ public Relation(RelationConstraint relation, IsaConstraint isa, Rule rule) {
             }
 
             @Override
-            public FunctionalIterator<Map<Identifier.Variable, Concept>> materialise(ConceptMap whenConcepts, TraversalEngine traversalEng,
-                                                                                     ConceptManager conceptMgr) {
-                Identifier.Variable relationTypeIdentifier = isa().type().id();
+            public Optional<Map<Identifier.Variable, Concept>> materialise(
+                    ConceptMap whenConcepts, TraversalEngine traversalEng, ConceptManager conceptMgr) {
                 RelationType relationType = relationType(whenConcepts, conceptMgr);
-                FunctionalIterator<com.vaticle.typedb.core.concept.thing.Relation> existingRelations = matchRelation(
-                        relationType, whenConcepts, traversalEng, conceptMgr
-                );
-                if (existingRelations.hasNext()) {
-                    return existingRelations.map(rel -> {
-                        Map<Identifier.Variable, Concept> thenConcepts = new HashMap<>();
-                        thenConcepts.put(relationTypeIdentifier, relationType);
-                        thenConcepts.put(isa().owner().id(), rel);
-                        relation().players().forEach(rp -> {
-                            thenConcepts.putIfAbsent(rp.roleType().get().id(), getRole(rp, relationType, whenConcepts));
-                            thenConcepts.putIfAbsent(rp.player().id(), whenConcepts.get(rp.player().id()));
-                        });
-                        return thenConcepts;
+                FunctionalIterator<com.vaticle.typedb.core.concept.thing.Relation> existingRelations =
+                        matchRelation(relationType, whenConcepts, traversalEng, conceptMgr);
+                if (!existingRelations.hasNext()) {
+                    return Optional.of(insert(relationType, whenConcepts));
+                } else {
+                    while (existingRelations.hasNext()) {
+                        com.vaticle.typedb.core.concept.thing.Relation preexisting = existingRelations.next();
+                        if (!preexisting.isInferred()) return Optional.empty();
+                        else {
+                            if (insertable(preexisting, whenConcepts)) {
+                                return Optional.of(thenConcepts(preexisting, whenConcepts));
+                            }
+                        }
+                    }
+                    return Optional.empty();
```

Comment:
I thought the same at first but there's a subtlety in the logic in line 485: the moment we find a non-inferred concept we return to avoid doing any extra work. Otherwise, if we ever find an inferred fact that was insertable then we need to return that materialisation. The fact that we need to break from the iterator prematurely for efficiency stops us from using `.filter().map().first()`. My assertion there is that it's helping exactly the fast relation lookup issue you have been fixing and so we don't want to sacrifice performance for code elegance in this instance.